pkg/lib/bundlev2/synthesis.go

```diff
@@ -0,0 +1,375 @@
+package bundlev2
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	log "github.com/sirupsen/logrus"
+	appsv1 "k8s.io/api/apps/v1"
+	cv1 "k8s.io/api/core/v1"
+	rv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/yaml"
+
+	"github.com/operator-framework/api/pkg/operators/v1alpha1"
+)
+
+const (
+	DefaultPermission = 0644
+	// MediaType
+	RegistryV2Type = "registry+v2"
+	PlainType      = "plain"
+	HelmType       = "helm"
+
+	AnnotationsFile = "annotations.yaml"
+	DockerFile      = "bundle.Dockerfile"
+	ManifestsDir    = "manifests/"
+	MetadataDir     = "metadata/"
+)
+
+type bundlev2 struct {
+	bundleDir           string
+	structuredObjectMap map[string]*[]interface{} // map[kind]: *array of structured objects
+	existingCSVPath     string
+	synthesisedCSV      *v1alpha1.ClusterServiceVersion
+	logger              *log.Entry
+}
+
+// NewBundleV2FromDirectory
+func NewBundleV2FromDirectory(bundleDir string) (*bundlev2, error) {
+	_, err := ioutil.ReadDir(bundleDir)
+	if err != nil {
+		return nil, err
+	}
+
+	return &bundlev2{
+		bundleDir:           bundleDir,
+		structuredObjectMap: map[string]*[]interface{}{},
+	}, nil
+}
+
+func (b *bundlev2) SynthesisedCSV() (*v1alpha1.ClusterServiceVersion, error) {
+	if err := b.loadStructuredObjects(); err != nil {
+		return nil, err
+	}
+
+	if err := b.synthesisCsv(); err != nil {
+		return nil, err
+	}
+
+	return b.synthesisedCSV, nil
+}
+
+func (b *bundlev2) synthesisCsv() error {
+	csvBasics, err := b.getRegistryV2CSV()
+	if err != nil {
+		return err
+	}
+
+	installStrategy, err := b.getInstallStrategy()
+	if err != nil {
+		return err
+	}
+
+	specCRDs, err := b.getSpecCRDs()
+	if err != nil {
+		return err
+	}
+
+	apiServiceDefs, err := b.getAPIServiceDefinitions()
+	if err != nil {
+		return err
+	}
+
+	nativeAPIs, err := b.getNativeAPIs()
+	if err != nil {
+		return err
+	}
+
+	b.synthesisedCSV = &v1alpha1.ClusterServiceVersion{
+		TypeMeta: v1.TypeMeta{
+			Kind:       CRDKind,
+			APIVersion: "operators.coreos.com/v1alpha1",
```

Comment:
nit: lets move this into a const or get this programatically somehow

Reply:
Yeah, i tried to find it. I'll try again, else i'll write it somewhere.